python/setup.py

```diff
@@ -149,35 +162,81 @@ def run(self):
       shutil.copy(source_path, destination_path)
 
 
-class InstallCommand(install):
+class BuildPyCommand(build_py.build_py):
   """Specialized Python builder to handle agent service dependencies.
 
-  During installation, this will comile the `agent_pb2{_grpc}.py` files and copy
+  During build, this will generate the `agent_pb2{_grpc}.py` files and copy
   `agent_service` binary next to `agent.py`.
   """
 
-  user_options = install.user_options
+  user_options = build_py.build_py.user_options
 
   def run(self):
-    self.run_command("compile_proto_grpc")
-    self.run_command("copy_agent_service_binary")
+    self.run_command("generate_proto_grpc")
     self.run_command("copy_task_assets")
-    install.run(self)
+    super().run()
 
 
-class DevelopCommand(develop):
-  """Specialized Python builder to handle agent service dependencies.
+class CMakeExtension(setuptools.Extension):
+  """A Python extension that has been prebuilt by CMake.
 
-  During installation, this will comile the `agent_pb2{_grpc,}.py` files and
-  copy `agent_service` binary next to `agent.py`.
+  We do not want distutils to handle the build process for our extensions, so
+  so we pass an empty list to the super constructor.
   """
 
-  user_options = develop.user_options
+  def __init__(self, name):
+    super().__init__(name, sources=[])
+
+
+class BuildCMakeExtension(build_ext.build_ext):
+  """Uses CMake to build extensions."""
 
   def run(self):
-    self.run_command("compile_proto_grpc")
+    self._configure_and_build_agent_service()
     self.run_command("copy_agent_service_binary")
-    install.run(self)
+
+  def _configure_and_build_agent_service(self):
+    """Check for CMake."""
+    cmake_command = "cmake"
+    build_cfg = "Debug"
+    mujoco_mpc_root = Path(__file__).parent.parent
+    mujoco_mpc_build_dir = mujoco_mpc_root / "build"
+    cmake_configure_args = [
+      "-DCMAKE_EXPORT_COMPILE_COMMANDS:BOOL=TRUE",
+      f"-DCMAKE_BUILD_TYPE:STRING={build_cfg}",
+      "-DBUILD_TESTING:BOOL=OFF"
+    ]
+
+    if platform.system() == "Darwin" and "ARCHFLAGS" in os.environ:
+      osx_archs = []
+      if "-arch x86_64" in os.environ["ARCHFLAGS"]:
+        osx_archs.append("x86_64")
+      if "-arch arm64" in os.environ["ARCHFLAGS"]:
+        osx_archs.append("arm64")
+      cmake_configure_args.append(
+        f"-DCMAKE_OSX_ARCHITECTURES={';'.join(osx_archs)}")
+
+    print("Configuring CMake with the following arguments:")
+    for arg in cmake_configure_args:
+      print(f"  {arg}")
+    subprocess.check_call([
+      cmake_command,
+      *cmake_configure_args,
+      f"-S{mujoco_mpc_root.resolve()}",
+      f"-B{mujoco_mpc_build_dir.resolve()}",
```

Comment:
This one currently configures the builds in `mujoco_mpc/build`. It might be better to use `self.build_temp` and `self.build_lib` for these but I thought it's probably fine this way, given that we currently anyway assume that the user has cloned the git repo.

Reply:
Done.